File: 496-next-greater-element-i/next-greater-element-i.py
class Solution(object):
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        stack = []
        ans = {}
        res = []
        for i in range(len(nums2) - 1, -1, -1):
            while stack and stack[-1] < nums2[i]:
                stack.pop()
            if not stack:
                ans[nums2[i]] = -1
            else:
                ans[nums2[i]] = stack[-1]
            stack.append(nums2[i])
        for i in nums1:
            res.append(ans[i])
        return res

File: 496-next-greater-element-i/next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> st = new Stack<>();
        // ArrayList<Integer> a = new ArrayList<>();
        HashMap<Integer, Integer> m = new HashMap<>();
        int a[] = new int[nums2.length];

        for (int i = nums2.length - 1; i>=0; i--) {
            m.put(nums2[i], i);
            while ((!st.isEmpty()) && nums2[i] >= st.peek()) {
                st.pop();
            }

            if (st.isEmpty()){
                a[i] = -1;
            }
            else {
                a[i] = st.peek();
            }

            st.push(nums2[i]);
        }

        // System.out.println(Arrays.toString(a) + " " + m);

        int[] arr = new int[nums1.length];
        for (int i = 0; i < nums1.length; i++) {
            arr[i] = (a[m.get(nums1[i])]);
        }
        return arr;
    }
}